[V1] Implement Cascade Attention

WoosukKwon opened the pull request:
This PR implements a simple version of [Cascade Attention](https://flashinfer.ai/2024/02/02/cascade-inference.html). Cascade attention can save the HBM bandwidth for reading KV cache when requests share the same prefix.

NOTE: For simplicity, this PR only uses cascade attention when every running request shares the same KV cache. If one or more requests do not share the KV cache, cascade attention is not used.

WoosukKwon committed:
[V1] Implement Cascade Inference

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

comaniac commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -271,6 +271,22 @@ def free(self, request: Request) -> None:
+        blocks = self.req_to_blocks[request.request_id]
+        num_common_blocks = 0
+        for block in blocks:
+            # FIXME(woosuk): For some reason, sometimes the ref_cnt is greater
+            # than the number of running requests. DEBUG this.
+            if block.ref_cnt >= num_requests:
+                num_common_blocks += 1
```
Please check https://github.com/vllm-project/vllm/pull/11565 and see if that solves the issue.

```suggestion
            if block.ref_cnt == len(self.running):
                num_common_blocks += 1
```

WoosukKwon replied:
Thanks for the bug fix. Fixed.

comaniac commented on vllm/v1/worker/gpu_model_runner.py:
```diff
@@ -355,6 +362,51 @@ def _prepare_inputs(self, scheduler_output: "SchedulerOutput"):
+            # Common case.
+            use_cascade = False
+        else:
+            # The common prefix should be already computed and stored in KV
+            # cache before this step.
```
Is this true for the first batch?

WoosukKwon replied:
Yes, this is required by the first kernel (computing prefix) in cascade attention.

For the first kernel, we treat all the query tokens (from different sequences) as a single sequence, and use bi-directional attention (causal=False) to compute the partial output. Therefore, every token in the query attends to the prefix without any masking. If the input query overlaps with the shared prefix in KV cache, it means we need to consider masking and can't use the bi-directional attention kernel.

WoosukKwon replied:
Actually, while writing this comment, I realized that we can relax the constraint a little bit: If the input query overlaps with the shared prefix in KV cache **only by 1 token**, the first kernel (computing prefix) is enough. Let me quickly prototype this.

WoosukKwon replied:
> if the input query overlaps with the shared prefix in KV cache only by 1 token, the first kernel (computing prefix) is enough.

After second thought, I found that this is a pretty niche case that probably isn't worth optimizing. The described case is essentially when all requests in the batch have the same query (of len 1) & same context in KV cache, which in turn means that the requests are duplicated. I think this is a very rare case.

Does this make sense to you? @comaniac

comaniac replied:
I guess I may just be confused by this comment "The common prefix should be already computed and stored in KV cache before this step". I suppose we call prefix and suffix consecutively in the following `cascade_attention` call, then why the common prefix should already be computed at this point?

WoosukKwon replied:
@comaniac Good point. My comment here was too vague.
I added a more detailed explanation with an example. PTAL!
```python
# NOTE(woosuk): Cascade attention uses two kernels: one for the
# common prefix and the other for the rest. For the first kernel,
# we concatenate all the query tokens (possibly from different
# requests) and treat them as if they are from a single request.
# Then, we use bi-directional attention to process the common prefix
# in the KV cache. Importantly, this means that the first kernel
# does not do any masking.

# Consider the following example:
# Request 1's input query: [D, E, X]
# Request 1's kv cache: [A, B, C, D, E, X]
# Request 1's num_computed_tokens: 3 (i.e., [A, B, C])
# Request 2's input query: [E, Y]
# Request 2's kv cache: [A, B, C, D, E, Y]
# Request 2's num_computed_tokens: 4 (i.e., [A, B, C, D])

# If we use [A, B, C, D, E] as the common prefix, then the
# first kernel will compute the bi-directional attention between
# input query [D, E, X, E, Y] and common prefix [A, B, C, D, E].
# However, this is wrong because D in Request 1 should not attend to
# E in the common prefix (i.e., we need masking).
# To avoid this, [A, B, C, D] should be the common prefix.
# That is, the common prefix should be capped by the minimum
# num_computed_tokens among the requests, and plus one to include
# the first token of the query.
common_prefix_len = min(
  common_prefix_len,
  self.input_batch.num_computed_tokens_cpu[:num_reqs].min() + 1)
```

WoosukKwon replied:
While writing this comment, I found that handling the prefix-only case is quite easy. I implemented it in https://github.com/vllm-project/vllm/pull/11635/commits/1dfd2d44a0ced15e0931f3a205918e45117af1e3

comaniac replied:
I see that makes sense. Also does that mean we cannot benefit from cascade attention in prefill? If we have 5 prefill requests with 90% of shared prompt in the same batch, we cannot benefit from this PR to save the computation?

WoosukKwon replied:
> does that mean we cannot benefit from cascade attention in prefill? If we have 5 prefill requests with 90% of shared prompt in the same batch, we cannot benefit from this PR to save the computation?

@comaniac We can save compute and KV cache storage thanks to prefix caching, regardless of cascade attention. Cascade attention is an additional optimization that saves HBM bandwidth in the attention computation. Without cascade attention, different requests are processed by different thread blocks, reading the shared KV cache redundantly. With cascade attention, we use the same thread blocks across multiple requests so that the shared KV cache is read less times. Again, it's just an additional optimization only affecting the performance of the attention kernel.

That being said, in the case you described, cascade attention will not be used due to the reason I explained above. Therefore, we will save computation and KV cache storage, but not HBM bandwidth for reading the KV cache during attention.

comaniac commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -169,21 +186,216 @@ def forward(
+    use_sliding_window: bool,
+    num_sms: int,
+) -> bool:
+    # Too short common prefix. Probably not worth using cascade attention.
+    # NOTE(woosuk): This is the common case. We should return False as soon as
+    # possible to avoid any unnecessary computation.
```
Yes it is always the case and this guard is pretty good

comaniac commented on vllm/v1/attention/backends/flash_attn.py:
```diff
@@ -169,21 +186,216 @@ def forward(
+    # Cascade attention is currently not supported with these variants.
+    if use_alibi or use_sliding_window:
+        return False
+    # Too few queries. Probably not worth using cascade attention.
+    num_reqs = len(query_lens)
+    if num_reqs < 8:
+        return False
```
I didn't think of this before but this is interesting. Worth evaluating further.

WoosukKwon replied:
This is actually a bit arbitrary number. Although it's theoretically faster than regular attention, cascade attention can be slow because it launches more kernels. Therefore, there will be a threshold that the benefit of cascade attention outweighs this overhead.

comaniac reviewed (commented):
Overall LGTM. The implementation is pretty clean. One optimization we can do in the future is having a wrapper CUDA kernel in flash attention so that we can save kernel invocation overheads. This is how FlashInfer does as well.

@raywanb since you've worked on FlashInfer cascade kernel integration in v0, can you also take a look at the current v1 interface and see if this interface can integrate FlashInfer easily in the future?

raywanb commented:
> Overall LGTM. The implementation is pretty clean. One optimization we can do in the future is having a wrapper CUDA kernel in flash attention so that we can save kernel invocation overheads. This is how FlashInfer does as well.
> 
> @raywanb since you've worked on FlashInfer cascade kernel integration in v0, can you also take a look at the current v1 interface and see if this interface can integrate FlashInfer easily in the future?

Yes, seems pretty straightforward to integrate Flashinfer!

WoosukKwon committed:
Merge branch 'main' into v1-cascade

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Merge branch 'main' into v1-cascade

WoosukKwon committed:
isort

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Comment

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -271,6 +271,22 @@ def free(self, request: Request) -> None:
+        self,
+        request: Request,
+        num_requests: int,
+    ) -> int:
```
Added a pretty detailed docstring (but didn't change the name since I felt it's a bit too long). PTAL.

comaniac commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -278,6 +278,54 @@ def free(self, request: Request) -> None:
+        1. The request has not yet finished, and 
+        2. The request holds its blocks unfreed.
+
+        While all scheduled requests must be in the RUNNING state, the inverse
+        is not necessarily true. There may be RUNNING requests that are not
+        scheduled in the current step. 
```
Could you elaborate on this? In what case a RUNNING request won't be scheduled in the current step? My understanding is if we cannot schedule all requests in the RUNNING queue, then we will preempt some of them. Preemption frees all their blocks and puts them back to the WAITING queue.

Also we have the following assertion. Doesn't it guarantee the number of scheduled requests is always equal to the number of RUNNING requests?

```
        assert (len(scheduled_new_reqs) + len(scheduled_resumed_reqs) +
                len(scheduled_running_reqs) == len(self.running))
```

WoosukKwon replied:
@comaniac You're right. In our current scheduler, all requests in the RUNNING state are scheduled every step. In other words, RUNNING requests are either scheduled or preempted (with the PREEMPTED state).

I wrote this comment because this might not be the case in the (near) future, where we would allow more flexibility in scheduling. I added the following to the docstring:
>  As of 1/1/2025, the scheduler does not allow this case, but it is possible in the future, as we allow more flexible scheduling.

comaniac reviewed (commented):
LGTM. Just two minor questions.
Also do you have a unit test to verify the correctness?

WoosukKwon committed:
comment

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
comment

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
docstring

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
comment

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Fix

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Consider prefix only

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
comment

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Minor

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Add debug

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Fix

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

WoosukKwon committed:
Add kernel test

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

comaniac reviewed (approved):
Thanks for adding this! I'll probably try to benchmark this feature next week if I got a chance.

WoosukKwon committed:
Add e2e test

Signed-off-by: Woosuk Kwon <woosuk.kwon@berkeley.edu>

sleepwalker2017 commented:
how does it work with cuda graph ? Seems they are not compatible?

comaniac commented:
> how does it work with cuda graph ? Seems they are not compatible?

v1 CUDA graph doesn't include attention layers so that we could easily implement attentions features like this without worrying about CUDA graph.

sleepwalker2017 commented:
> > how does it work with cuda graph ? Seems they are not compatible?
> 
> v1 CUDA graph doesn't include attention layers so that we could easily implement attentions features like this without worrying about CUDA graph.

wow, that's a big change, I'll learn about it. Thank you!

sleepwalker2017 commented:
> > how does it work with cuda graph ? Seems they are not compatible?
> 
> v1 CUDA graph doesn't include attention layers so that we could easily implement attentions features like this without worrying about CUDA graph.

I still have a question, look at the gap between cascade kernels, does it really have benefits over one large kernel? 
![image](https://github.com/user-attachments/assets/ccfc1fd5-1c79-4834-aefc-e25c321eb2f8)

WoosukKwon commented:
@sleepwalker2017 Thanks for trying it out! Your overhead looks much larger than I expected: not only for cascade attention, but also the "gap" between other kernels looks so bad. Could you please provide your setup (GPU, model, batch size, prefix length, etc.)?

sleepwalker2017 commented:
> @sleepwalker2017 Thanks for trying it out! Your overhead looks much larger than I expected: not only for cascade attention, but also the "gap" between other kernels looks so bad. Could you please provide your setup (GPU, model, batch size, prefix length, etc.)?

I run it in eager mode.  

My GPU: A100, 

My code here:

cmd : 
```
VLLM_USE_V1=1 VLLM_ATTENTION_BACKEND=FLASHINFER \
nsys profile --trace-fork-before-exec=true --cuda-graph-trace=node --duration 60 --force-overwrite true \
python offline.py

```

```

from vllm import LLM, SamplingParams

# Sample prompts.
prompts = [
    "how old are you",
    "Hello, my name is Jack." * 100 + '1',
    "Hello, my name is Jack." * 100 + '2',
    "Hello, my name is Jack." * 100 + '3',
    "Hello, my name is Jack." * 100 + '4',
    "Hello, my name is Jack." * 100 + '5',
    "Hello, my name is Jack." * 100 + '6',
    "Hello, my name is Jack." * 100 + '7',
    "Hello, my name is Jack." * 100 + '8',
]
# Create a sampling params object.
sampling_params = SamplingParams(temperature=0.8, top_p=0.95, max_tokens=3)

# Create an LLM.
llm = LLM(model="/data/models/Llama-2-7b-hf", enforce_eager=not True)
# Generate texts from the prompts. The output is a list of RequestOutput objects
# that contain the prompt, generated text, and other information.

#import pdb; pdb.set_trace()
import nvtx
nvtx.push_range("generate")
outputs = llm.generate(prompts, sampling_params)
nvtx.pop_range()
# Print the outputs.
print("Done")
while True:
    #break
    pass
```

BTW, I'm very confused about the piece wise cuda graph, and I can't find the code that really captures the kernels.

I trace the code here:
```
@contextmanager
def graph_capture(device: torch.device):
    """
    `graph_capture` is a context manager which should surround the code that
    is capturing the CUDA graph. Its main purpose is to ensure that the
    some operations will be run after the graph is captured, before the graph
    is replayed. It returns a `GraphCaptureContext` object which contains the
    necessary data for the graph capture. Currently, it only contains the
    stream that the graph capture is running on. This stream is set to the
    current CUDA stream when the context manager is entered and reset to the
    default stream when the context manager is exited. This is to ensure that
    the graph capture is running on a separate stream from the default stream,
    in order to explicitly distinguish the kernels to capture
    from other kernels possibly launched on background in the default stream.
    """
    context = GraphCaptureContext(torch.cuda.Stream(device=device))
    with get_tp_group().graph_capture(context), get_pp_group().graph_capture(
            context):
        yield context
```

It just gives a cuda stream to run the kernels, but I can't find the torch.cuda.Graph object.

Could you give some instructions about this feature?

sleepwalker2017 commented:
> > @sleepwalker2017 Thanks for trying it out! Your overhead looks much larger than I expected: not only for cascade attention, but also the "gap" between other kernels looks so bad. Could you please provide your setup (GPU, model, batch size, prefix length, etc.)?
> 
> I run it in eager mode.
> 
> My GPU: A100, vllm v0.7.1, because I have some problems to run 0.7.2.
> 
> My code here:
> 
> cmd :
> 
> ```
> VLLM_USE_V1=1 \
> nsys profile --trace-fork-before-exec=true --cuda-graph-trace=node --duration 60 --force-overwrite true \
> python offline.py
> ```
> 
> ```
> 
> from vllm import LLM, SamplingParams
> 
> # Sample prompts.
> prompts = [
>     "how old are you",
>     "Hello, my name is Jack." * 100 + '1',
>     "Hello, my name is Jack." * 100 + '2',
>     "Hello, my name is Jack." * 100 + '3',
>     "Hello, my name is Jack." * 100 + '4',
>     "Hello, my name is Jack." * 100 + '5',
>     "Hello, my name is Jack." * 100 + '6',
>     "Hello, my name is Jack." * 100 + '7',
>     "Hello, my name is Jack." * 100 + '8',
> ]
> # Create a sampling params object.
> sampling_params = SamplingParams(temperature=0.8, top_p=0.95, max_tokens=3)
> 
> # Create an LLM.
> llm = LLM(model="/data/models/Llama-2-7b-hf", enforce_eager=True)
> # Generate texts from the prompts. The output is a list of RequestOutput objects
> # that contain the prompt, generated text, and other information.
> 
> #import pdb; pdb.set_trace()
> import nvtx
> nvtx.push_range("generate")
> outputs = llm.generate(prompts, sampling_params)
> nvtx.pop_range()
> # Print the outputs.
> print("Done")
> while True:
>     #break
>     pass
> ```
> 
> BTW, I'm very confused about the piece wise cuda graph, and I can't find the code that really captures the kernels.
> 
> I trace the code here:
> 
> ```
> @contextmanager
> def graph_capture(device: torch.device):
>     """
>     `graph_capture` is a context manager which should surround the code that
>     is capturing the CUDA graph. Its main purpose is to ensure that the
>     some operations will be run after the graph is captured, before the graph
>     is replayed. It returns a `GraphCaptureContext` object which contains the
>     necessary data for the graph capture. Currently, it only contains the
>     stream that the graph capture is running on. This stream is set to the
>     current CUDA stream when the context manager is entered and reset to the
>     default stream when the context manager is exited. This is to ensure that
>     the graph capture is running on a separate stream from the default stream,
>     in order to explicitly distinguish the kernels to capture
>     from other kernels possibly launched on background in the default stream.
>     """
>     context = GraphCaptureContext(torch.cuda.Stream(device=device))
>     with get_tp_group().graph_capture(context), get_pp_group().graph_capture(
>             context):
>         yield context
> ```
> 
> It just gives a cuda stream to run the kernels, but I can't find the torch.cuda.Graph object.
> 
> Could you give some instructions about this feature?

I can't find `CUDAGraph` keywords in V1 implementation. That's so different from the V0 version.

WoosukKwon commented:
@sleepwalker2017 I see. Please do not use `enfoce_eager`!

In V1, we introduced [piecewise CUDA graphs](https://docs.google.com/presentation/d/1epVkt4Zu8Jz_S5OhEHPc798emsYh2BwYfRuDDVEF7u4/edit?usp=sharing) which captures all ops except attention and runs the attention op in eager PyTorch. This gives a lot of freedom for us to implement/optimize the attention op while retaining most of the performance gain from CUDA graphs.

We use `torch.compile` for this. We implemented a compiler pass that detects the attention ops in the model, splits the graph, and generates CUDA graphs for each "piece" (subgraphs) before and after attention. That's why you don't see explicit `CUDAGraph` objects in our code.

sleepwalker2017 commented:
> @sleepwalker2017 I see. Please do not use `enfoce_eager`!
> 
> In V1, we introduced [piecewise CUDA graphs](https://docs.google.com/presentation/d/1epVkt4Zu8Jz_S5OhEHPc798emsYh2BwYfRuDDVEF7u4/edit?usp=sharing) which captures all ops except attention and runs the attention op in eager PyTorch. This gives a lot of freedom for us to implement/optimize the attention op while retaining most of the performance gain from CUDA graphs.
> 
> We use `torch.compile` for this. We implemented a compiler pass that detects the attention ops in the model, splits the graph, and generates CUDA graphs for each "piece" (subgraphs) before and after attention. That's why you don't see explicit `CUDAGraph` objects in our code.

Thank you!  I'll read the slides.  
It seems we can't add nvtx annotations in non-eager mode? I try to profile these kernels, but failed in non-eager mode.
